### DISCUSSION TITLE
DANALOCK and Pin code support

### DISCUSSION THREAD

**Discussion by XLDom** [Koenkk/zigbee2mqtt#12706] [Devices]
Hello @Koenkk,

why this rollback https://github.com/Koenkk/zigbee-herdsman-converters/commit/5177b7749a47f448e3c26a7b0bffa51112716cba ?

i've the Danalock and i've push request for pin_code.
Pin_code is supported and run well

Can you please cancel this remove upgrade :)

Regards

**Comment by Koenkk** [comment]
See https://github.com/Koenkk/zigbee-herdsman-converters/pull/3665#issuecomment-1128784604

**Reply by XLDom** [reply]
@Koenkk i've see this post.... I also spoke with Danalock team who send me the doc.
It was after that that I made a push request to add this feature which works very well...

Just see the Danalock official documentation page 12 please...
and I repeat myself, but it works perfectly. :)

[MAN003 DanalockV3 Zigbee Product Manual - 20211209.pdf](https://github.com/Koenkk/zigbee2mqtt/files/8840059/MAN003.DanalockV3.Zigbee.Product.Manual.-.20211209.pdf)

**Reply by Koenkk** [reply]
FYI @smulle48, will add it back.

**Reply by smulle48** [reply]
Well. As before, it still don't work. My entire zigbee network crash when trying to add an pincode and the only way to get the lock working again, is to reset it, delete it and add it again.

It don't even send pincode, when locking and unlocking with the keypad.

Debug 2022-06-05 22:30:33Received Zigbee message from 'Hoveddør Lås', type 'commandOperationEventNotification', cluster 'closuresDoorLock', data '{"data":0,"opereventcode":2,"opereventsrc":1,"pin":{"data":[],"type":"Buffer"},"userid":65535,"zigbeelocaltime":707775715}' from endpoint 1 with groupID 0
Info 2022-06-05 22:30:33MQTT publish: topic 'zigbee2mqtt/Hoveddør Lås', payload '{"action":"unlock","action_source":1,"action_source_name":"rf","action_source_user":null,"action_user":65535,"battery":80,"linkquality":72,"lock_state":"locked","state":"LOCK"}'
Info 2022-06-05 22:30:33MQTT publish: topic 'zigbee2mqtt/Hoveddør Lås', payload '{"action":"","action_source_name":null,"action_source_user":null,"battery":80,"linkquality":72,"lock_state":"locked","state":"LOCK"}'
Info 2022-06-05 22:30:33MQTT publish: topic 'zigbee2mqtt/Hoveddør Lås/action', payload 'unlock'
Debug 2022-06-05 22:30:33Received Zigbee message from 'Hoveddør Lås', type 'attributeReport', cluster 'closuresDoorLock', data '{"lockState":2}' from endpoint 1 with groupID 0
Info 2022-06-05 22:30:33MQTT publish: topic 'zigbee2mqtt/Hoveddør Lås', payload '{"action":null,"action_source_name":null,"action_source_user":null,"battery":80,"linkquality":72,"lock_state":"unlocked","state":"UNLOCK"}'


Debug 2022-06-05 22:40:35Received MQTT message on 'zigbee2mqtt/Hoveddør Lås/set' with data '{"pin_code":{"user":2,"pin_code":1234,"user_type":"unrestricted","user_enabled":true},"user_status":{"user":0,"status":"enabled"}}'
Debug 2022-06-05 22:40:35Publishing 'set' 'pin_code' to 'Hoveddør Lås'
Debug 2022-06-05 22:40:59Received Zigbee message from 'Hoveddør Lås', type 'attributeReport', cluster 'closuresDoorLock', data '{"lockState":0}' from endpoint 1 with groupID 0
Info 2022-06-05 22:40:59MQTT publish: topic 'zigbee2mqtt/Hoveddør Lås', payload '{"action":null,"action_source_name":null,"action_source_user":null,"battery":80,"linkquality":80,"lock_state":"not_fully_locked","state":"UNLOCK"}'
Error 2022-06-05 22:41:22Publish 'set' 'pin_code' to 'Hoveddør Lås' failed: 'Error: Command 0x000b57fffe59e37d/1 closuresDoorLock.setPinCode({"userid":2,"userstatus":1,"usertype":0,"pincodevalue":"1234"}, {"sendWhen":"active","timeout":10000,"disableResponse":false,"disableRecovery":false,"disableDefaultResponse":true,"direction":0,"srcEndpoint":null,"reservedBits":0,"manufacturerCode":null,"transactionSequenceNumber":null,"writeUndiv":false}) failed (Timeout - 65161 - 1 - 81 - 257 - 5 after 10000ms)'

@XLDom What firmware are you running on the lock? Because danalock told me that it was not supported on fw 0.20.0 due to some bugs.
![Skærmbillede 2022-06-05 kl  22 49 47](https://user-images.githubusercontent.com/24297824/172070030-2d957945-dea3-4d07-abf2-6fa767a34116.png)

![Skærmbillede 2022-06-05 kl  22 49 55](https://user-images.githubusercontent.com/24297824/172070036-aad58dee-487c-4da9-9e1c-1ade01407403.png)

![Skærmbillede 2022-06-05 kl  22 50 02](https://user-images.githubusercontent.com/24297824/172070047-5d45efd3-94ea-4704-8312-821d87e7bf34.png)

![Skærmbillede 2022-06-05 kl  22 50 08](https://user-images.githubusercontent.com/24297824/172070057-2db7e7f9-7895-441e-bf86-89da4b4b5869.png)

![Skærmbillede 2022-06-05 kl  22 50 16](https://user-images.githubusercontent.com/24297824/172070065-bd0f9c52-aa6b-4366-9b87-fbb530aa9d15.png)

**Comment by Koenkk** [comment]
Added it back.

Changes will be available in the dev branch in a few hours from now. (https://www.zigbee2mqtt.io/advanced/more/switch-to-dev-branch.html)

**Reply by XLDom** [reply]
Thx @Koenkk,

@smulle48 What is your adapter key ? have you the last firmware on ?
And for the danalock, why don't you use the join menu on the danalock app to repair ?
Other question, have you try to just remove and place a battery on the danalock and see if she work ?

**Reply by smulle48** [reply]
I'm using the Slae.sh with the latest firmware.

I have tried multiple times to repair. After the first attempt to add a pincode to the lock, the connection dies and I have to start all over.

Yes, I have tried to remove the batteries and insert them again, but no success

**Reply by smulle48** [reply]
I have tried with a spare conbee2, but with same result.

**Reply by smulle48** [reply]
ConBee (Dresden) is also experiencing problems with the latest firmware, and Danalock has acknowledged that there is indeed an bug in the firmware :-(

https://forum.phoscon.de/t/danalock-v3-new-fw-0-20-0-losing-connection/1838/6

**Reply by XLDom** [reply]
Just write to support to have old firmware file (you can inject with "Danalock toolbox" app i think)
Tell me if it's ok :)

**Reply by smulle48** [reply]
> Just write to support to have old firmware file (you can inject with "Danalock toolbox" app i think)
> 
> Tell me if it's ok :)

They don't want to give me the older firmware file 🤬

Guess I have to wait for them to fix it, or just give up and get an decent lock. There has been nothing but problems with the zigbee version

**Reply by XLDom** [reply]
@smulle48 Danalock released a new firmware. Can you test please, with dev version of zigbee2mqtt

**Reply by smulle48** [reply]
> @smulle48 Danalock released a new firmware. Can you test please, with dev version of zigbee2mqtt

Have tested it the last 24 hours and it's even more cra**y the the last firmware.

The lock still don't accept pincodes, nor does it expose the pin in the log.

Worst of all is that now the zigbee signal gets killed after about an hour. It simply goes into deep sleep and won't report anything. Even manually turning the lock does nothing. Only solution is to remove the batteries. That will wake up the lock for another hour or so.

Danalock won't provide me with the old firmware 🤬🤬

So do yourself a favour and DONT update. You will regret it 😞

**Reply by smulle48** [reply]
@XLDom successfully managed to downgrade my lock to 0.17.15 and now the lock is stable again, but still my lock won't accept creating pin code, nor will it post pin code in the log output :-(

**Reply by smulle48** [reply]
Just an update.

I'm waiting for the final confirmation, but it seems like my lock is too old to support all features. There has been an hw revision, and apparently the old model don't support it.

The good thing is that the old lock don't got a locked boot loader. That's why I'm able to downgrade my lock, even when Danalock says it's not possible 😃

I'll report back as soon as I get confirmation. If true, the code might need some tweaking to distinguish between the old and the new revision.

**Reply by XLDom** [reply]
Thanks for all information !

you have rollback your firmware ?

**Reply by smulle48** [reply]
@XLDom 

Finally got an answer from the developers.

According to them, the ZigBee module in my lock is only Zigbee 1.2 that don't support pincode. The newer (locked bootloader) is ZigBee 3.0 which does support pincode.

As my lock is the old ZigBee 1.2 the boot loader is not locked, so I can upgrade and downgrade the firmware as I want. This is not possible on the newer ZigBee 3.0 as the boot loader is locked, preventing downgrade.

Danalock was kind enough to upgrade my lock to ZigBee 3.0 for $40. I know it's not cheap for such a little change, but at least cheaper then buying a new lock.

I propose an "fix" to the code to distinguish between the old and the new version to either show or hide pincode support. I don't know if the firmware build date can be used for that? or is it possible to pull the ZigBee version from the lock?

**Reply by smulle48** [reply]
@XLDom @Koenkk just got my lock back from Danalock. I can confirm that all the functions ARE working now. So it's correct that the pin code only works on locks that are Zigbee3.0

The older locks that are Zigbee1.2 does not support it (but they can downgrade their firmware)
